subprojects/model-core/src/integTest/groovy/org/gradle/api/provider/PropertyAssignmentIntegrationTest.groovy

```diff
@@ -0,0 +1,443 @@
+        buildKotlinFile.text = KOTLIN_BUILD_FILE_TEMPLATE
+            .replace("{inputDeclaration}", "abstract val input: $inputType")
+            .replace("{inputValue}", inputValue)
+            .replace("{operation}", operation)
+
+        expect:
+        runAndAssert("myTask", expectedResult)
+
+        where:
+        description                        | operation | inputType                    | inputValue                         | expectedResult
+        "FileCollection = FileCollection"  | "="       | "ConfigurableFileCollection" | 'files("a.txt") as FileCollection' | unsupportedWithDescription("Val cannot be reassigned")
+        "FileCollection = String"          | "="       | "ConfigurableFileCollection" | '"a.txt"'                          | unsupportedWithDescription("Val cannot be reassigned")
+        "FileCollection = Object"          | "="       | "ConfigurableFileCollection" | 'MyObject("a.txt")'                | unsupportedWithDescription("Val cannot be reassigned")
+        "FileCollection = File"            | "="       | "ConfigurableFileCollection" | 'file("a.txt")'                    | unsupportedWithDescription("Val cannot be reassigned")
+        "FileCollection = Iterable<File>"  | "="       | "ConfigurableFileCollection" | 'listOf(file("a.txt"))'            | unsupportedWithDescription("Val cannot be reassigned")
+        "FileCollection += FileCollection" | "+="      | "ConfigurableFileCollection" | 'files("a.txt") as FileCollection' | unsupportedWithDescription("Val cannot be reassigned")
+        "FileCollection += String"         | "+="      | "ConfigurableFileCollection" | '"a.txt"'                          | unsupportedWithDescription("Val cannot be reassigned")
+        "FileCollection += Object"         | "+="      | "ConfigurableFileCollection" | 'MyObject("a.txt")'                | unsupportedWithDescription("Val cannot be reassigned")
+        "FileCollection += File"           | "+="      | "ConfigurableFileCollection" | 'file("a.txt")'                    | unsupportedWithDescription("Val cannot be reassigned")
+        "FileCollection += Iterable<?>"    | "+="      | "ConfigurableFileCollection" | 'listOf("a.txt")'                  | unsupportedWithDescription("Val cannot be reassigned")
+        "FileCollection += Iterable<File>" | "+="      | "ConfigurableFileCollection" | 'listOf(file("a.txt"))'            | unsupportedWithDescription("Val cannot be reassigned")
+    }
+
+    private void runAndAssert(String task, Object expectedResult) {
+        if (expectedResult instanceof FailureWithCause) {
+            def failure = runAndFail(task)
+            failure.assertHasCause(expectedResult.failureCause)
+        } else if (expectedResult instanceof FailureWithDescription) {
+            def failure = runAndFail(task)
+            failure.assertThatDescription(containsNormalizedString(expectedResult.failureDescription))
+        } else {
```

Comment:
🤔 I think it would be somewhat nicer to do the checks via an method implemented in `FailureWithCause` and `FailureWithDescription`. You could define a `Failure` interface with the method, and then you only need to have one `instanceof` call here.

Reply:
We can also go with that.